cmd2.py

```diff
@@ -969,25 +970,47 @@ def _surround_ansi_escapes(prompt, start="\x01", end="\x02"):
         return result
 
     def pseudo_raw_input(self, prompt):
-        """copied from cmd's cmdloop; like raw_input, but accounts for changed stdin, stdout"""
+        """
+        began life as a copy of cmd's cmdloop; like raw_input but
+        
+        - accounts for changed stdin, stdout
+        - if input is a pipe (instead of a tty), look at self.echo
+          to decide whether to print the prompt and the input
+        """
 
         # Deal with the vagaries of readline and ANSI escape codes
         safe_prompt = self._surround_ansi_escapes(prompt)
 
         if self.use_rawinput:
             try:
-                line = sm.input(safe_prompt)
+                if sys.stdin.isatty():
+                    sys.stdout.write(safe_prompt)
```

Comment:
My (apparently lame) thought was that I can patch or mock the **sm.input()** but since I don't know how to get the side effect of echoing the characters to stdout, I'd just print them out there first. But if it breaks real-life, who cares whether the tests pass or are easy?